### PR TITLE
Add payload conversion and heap measurement tool

### LINKED ISSUE #1661: Write tooling to keep track of (and enforce) that our data structs do not allocate
Part of https://github.com/unicode-org/icu4x/issues/856

We should keep track of which keys are not zero-copy yet, via a test that uses [`assert_no_alloc`](https://docs.rs/assert_no_alloc/) or [`dhat`](https://docs.rs/dhat), and store the list in a sorted golden file so that any changes get tracked.

cc @sffc @robertbastian

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/unicode-org/icu4x/issues/1512, fixes #1661

This adds `icu4x-verify-zero-copy`, that loads `testdata.postcard` and deserializes each entry that it knows to look for to its specific data struct type, while measuring heap usage. It then reports which keys have non-zero-copy data structs. (this is #1661)

This also updates `impl_dyn_provider!()` to be able to provide `DataConverter` impls that act as a "data registry" (#1512), allowing one to convert between different dynamic marker types via intermediate downcasting.

Todo:
 - [x] docs and comments
 - [x] run this in CI

### CODE REVIEW

**Comment by Manishearth** [comment]
cc @robertbastian @zbraniecki

**Comment by Manishearth** [comment]
cc @robertbastian btw

File: tools/datagen/src/bin/verify-zero-copy.rs [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                .help("Path to the uprops directory.")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("INPUT_FROM_TESTDATA")
+                .long("input-from-testdata")
+                .help("Load input data from the icu_testdata project."),
+        )
+        .arg(
+            Arg::with_name("KEYS")
```
Comment (Manishearth) [inline tools/datagen/src/bin/verify-zero-copy.rs:79]: I added this because I found it useful to be `--keys specifickey -v` to debug things

File: provider/core/src/dynutil.rs [hunk 2]
```diff
@@ -189,6 +204,52 @@ macro_rules! impl_dyn_provider {
...
+    };
+    ($provider:ty, { $($pat:pat $(if $guard:expr)? => $struct_m:ty),+, }, impl DataConverter) => {
+        // If this fails to compile, enable the "datagen" feature on this crate.
+        impl $crate::datagen::DataConverter<$crate::any::AnyMarker, $crate::serde::SerializeMarker> for $provider {
+            fn convert(&self, key: $crate::ResourceKey, from: DataPayload<$crate::any::AnyMarker>) -> Result<$crate::DataPayload<$crate::serde::SerializeMarker>, $crate::datagen::ReturnedPayloadError<$crate::any::AnyMarker>> {
+                use $crate::datagen::ReturnedPayloadError;
+                match key {
+                    $(
+                        $pat $(if $guard)? => {
+                            let result: $crate::DataPayload<$struct_m> = from.downcast().expect("Downcasting failed!");
```
Comment (sffc) [inline provider/core/src/dynutil.rs:229]: Nit: Fix this to return a `ReturnedPayloadError` populated with the payload that failed to downcast

File: provider/core/src/datagen/heap_measure.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use yoke::trait_hack::YokeTraitHack;
+
+/// Stats on the heap size needed when attempting to zero-copy-deserialize
+/// a postcard-formatted data struct.
+#[derive(Debug, Copy, Clone, yoke::Yokeable, Default)]
+pub struct HeapStats {
+    // Total bytes allocated during deserialization
+    pub total_bytes_allocated: u64,
+    // Total bytes allocated during deserialization that have not yet been freed
+    pub final_bytes_needed: usize,
```
Comment (sffc) [inline provider/core/src/datagen/heap_measure.rs:16]: Suggestion (optional):
```suggestion
    pub net_bytes_allocated: usize,
```

File: provider/core/src/datagen/data_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    fn convert(
+        &self,
+        key: crate::ResourceKey,
+        from: DataPayload<MFrom>,
+    ) -> Result<DataPayload<MTo>, ReturnedPayloadError<MFrom>> {
+        (**self).convert(key, from)
+    }
+}
+
+pub struct ReturnedPayloadError<M: DataMarker>(pub DataPayload<M>, pub DataError);
```
Comment (sffc) [inline provider/core/src/datagen/data_conversion.rs:43]: It looks like you added this `ReturnedPayloadError` but didn't tie it in with the actual `downcast()` function.

I suggest moving this into the `any` module and then returning it from `downcast()`.

File: tools/datagen/src/bin/datagen.rs.orig [hunk 5]
```diff
@@ -0,0 +1,513 @@
+// This file is part of ICU4X. For terms of use, please see the file
```
Comment (sffc) [inline tools/datagen/src/bin/datagen.rs.orig:1]: Nit: Remove this file

**Review by sffc — CHANGES_REQUESTED** [review]
First batch; more in the morning

File: provider/core/src/datagen/data_conversion.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+/// concrete marker types M, and reifying the input to a `DataPayload<M>`, performing some conversion
+/// or computation, and erasing the result to `DataPayload<MTo>`.
+///
+/// It will typically be implemented on data providers used in datagen.
+///
+/// The `impl_dyn_provider` is able to automatically implement this trait.
+pub trait DataConverter<MFrom: DataMarker, MTo: DataMarker> {
+    fn convert(
+        &self,
+        key: crate::ResourceKey,
```
Comment (robertbastian) [inline provider/core/src/datagen/data_conversion.rs:23]: Nit: import `ResourceKey` as it's referenced multiple times

File: provider/core/src/datagen/heap_measure.rs (hunk 3 above)
Comment (robertbastian) [reply provider/core/src/datagen/heap_measure.rs:16]: -1, I think `final` is clearer than `net`. Maybe `intermediate` (`total-final`) and `final`?

File: provider/core/src/datagen/mod.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+use crate::any::AnyMarker;
+use crate::buf::BufferMarker;
+use crate::iter::IterableDynProvider;
+use crate::serde::SerializeMarker;
+use crate::DataMarker;
+
+/// A blanket-implemented trait that allows for all the datagen-relevant traits
+/// to be used in trait objects. The type parameter is the marker type needed for
+/// [`IterableDynProvider`].
+pub trait OmnibusDatagenProvider<M: DataMarker>:
```
Comment (robertbastian) [inline provider/core/src/datagen/mod.rs:26]: Can we just call this `DatagenProvider`?

Actually in #1092 we said that `IterableDynProvider` is not the right trait for runtime, so maybe we can just rename that `DatagenProvider` and make it extend the `DataConverter` impls? It should also be behind the `datagen` feature now that that exists.

File: provider/core/src/dynutil.rs [hunk 8]
```diff
@@ -189,6 +204,52 @@ macro_rules! impl_dyn_provider {
             $crate::serde::SerializeMarker
         );
     };
+    ($provider:ty, [ $($struct_m:ty),+, ], SERDE_SE, impl DataConverter) => {
```
Comment (robertbastian) [inline provider/core/src/dynutil.rs:207]: This macro is getting massive. Now that we have a `datagen` feature we could split it into what it was before #1595, and a second macro (`impl_datagen_provider`) that additionally implements `IterableDynProvider` and `DataConverter` (and which we can also put behind the `datagen` feature).

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 1 above)
Comment (robertbastian) [reply tools/datagen/src/bin/verify-zero-copy.rs:83]: Also add `ALL_KEYS`?

File: tools/datagen/src/bin/verify-zero-copy.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    });
+
+    let converter: Box<dyn OmnibusDatagenProvider<SerializeMarker> + Sync> =
+        Box::new(ForkByKeyProvider(
+            icu_provider_cldr::create_exportable_provider(
+                cldr_paths.as_ref(),
+                uprops_root.clone(),
+            )?,
+            icu_provider_uprops::create_exportable_provider(&uprops_root)?,
+        ));
```
Comment (robertbastian) [inline tools/datagen/src/bin/verify-zero-copy.rs:171]: I don't like the `HeapStatsMarker`/`convert` design. The raw CLDR/uprops data should not be required by this test, and yet it is. Also it adds the benchmarking code and dependency in the `icu_provider` crate, even if it's behind feature, I think it should really all be in this crate.

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: provider/core/src/dynutil.rs (hunk 2 above)
Comment (Manishearth) [reply provider/core/src/dynutil.rs:229]: This needs the downcasting methods to also do the same everywhere.

There are a bunch of places ReturnedPayloadError needs to be introduced, perhaps we can do it all at once in a separate PR?

Given that we're already checking the key type it's somewhat of a code bug if this branch is hit anyway so the expect is also somewhat correct.

File: provider/core/src/datagen/data_conversion.rs (hunk 4 above)
Comment (Manishearth) [reply provider/core/src/datagen/data_conversion.rs:43]: Yes, the plan was to do it later, it's not just any that needs it.

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: I'd rather do wider refactors around what's under what feature as a result of https://github.com/unicode-org/icu4x/issues/1320 (which I plan to get done soon)

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 1 above)
Comment (Manishearth) [reply tools/datagen/src/bin/verify-zero-copy.rs:83]: There's currently no difference and we can add that as needed I think

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 9 above)
Comment (sffc) [reply tools/datagen/src/bin/verify-zero-copy.rs:171]: About the CLDR/uprops data: it is not actually required for anything; it is only here because the impl for DataConverter is on CldrJsonDataprovider and the various PropertiesProviders, and the only way to get that impl is to construct them. Better would be if we could get a "static" instance, since all we need is the type info.

About the location of the benchmarking code: the other way to do it would be to have a callback function. It gets rid of the optional dehat dependency in icu_provider, which is nice. But I like the DataConverter design too (Manish and I designed it on a whiteboard). If you think you can do better with a callback function, you could make an attempt after this PR is landed (after finishing the OKR goals).

Comment (Manishearth) [reply tools/datagen/src/bin/verify-zero-copy.rs:171]: It can't be since the registry code needs to be everywhere. I did have a registry design that was more generic but it had some problems and we opted for this which is more modular (and allows for the crab bake registry, for example, to not handle all types)

You're right that the raw data isn't super needed here. We can potentially add modes to these providers that choose to not load data but I'm unsure if we should.

Overall the `convert()` design is needed for a bunch of other things too, that's why we picked it.

File: tools/datagen/src/bin/verify-zero-copy.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+    // Litemap keeps it sorted, convenient
+    let mut violations: LiteMap<&'static str, u64> = LiteMap::new();
+
+    for key in selected_keys.into_iter() {
+        let props_key = key.get_path().starts_with("props/");
+
+        let mut max_violation = 0;
+
+        for options in converter.supported_options_for_key(key)? {
```
Comment (sffc) [inline tools/datagen/src/bin/verify-zero-copy.rs:200]: Suggestion: Actually it is kind-of odd to use the supported_options from the source data and then load it from the postcard file. I'm warming up to the idea that we should read from source data, serialize, and deserialize. This can be in another PR.

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (robertbastian) [reply provider/core/src/datagen/mod.rs:26]: Still, what do buses have to do with this

File: provider/core/src/dynutil.rs (hunk 2 above)
Comment (sffc) [reply provider/core/src/dynutil.rs:229]: OK to do it in another PR.

File: provider/core/src/datagen/heap_measure.rs (hunk 3 above)
Comment (Manishearth) [reply provider/core/src/datagen/heap_measure.rs]: Yeah, I agree with "net"

Comment (robertbastian) [reply provider/core/src/datagen/heap_measure.rs]: `final` is the final number of bytes that are allocated for this struct. `intermediate` is the number of additional bytes that were allocated during deserialization but then freed again. The fact that we computed these numbers by looking at total memory usage before and after should not have any impact on the naming.

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: "Omnibus" is just a term used in English to talk about things that contain more things than usual, typically in the case of media around anthologies and "3-in-1 books".

In this case, this is really just "a lot of traits in one trait", hence the trait name.

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 10 above)
Comment (Manishearth) [reply tools/datagen/src/bin/verify-zero-copy.rs:200]: The idea was that (a) the blob provider doesn't know about supported options and (b) i'd like to check for all options that are _supposed_ to be in the file.

**Comment by Manishearth** [comment]
Rebased over segmenter changes, no changes between your last review and this except for https://github.com/unicode-org/icu4x/pull/1670/commits/4c8b20838dc53acbaf143984f8865419eb7fba2f and merge conflict resolution

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (robertbastian) [reply provider/core/src/datagen/mod.rs:26]: I have never seen it used that way, maybe it's an American thing. I'd prefer if we avoid such words, we should keep our code accessible to non-Americans and non-native English speakers.

File: provider/core/src/dynutil.rs (hunk 8 above)
Comment (Manishearth) [reply provider/core/src/dynutil.rs:207]: I'm waiting for the ResourceProvider stuff to happen before making further changes here

**Review by sffc — APPROVED** [review]

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: Hmm, so I kinda prefer the longer name because we have a lot of FooProvider types already and it gets confusing, i think it would be useful to have a longer name here which is why I picked Omnibus. I could use AllInOne but every time I see AllInOne I just feel like Omnibus is a better name there.

Comment (sffc) [reply provider/core/src/datagen/mod.rs:26]: Does this need to be a public type, or could it be internal to the datagen crate? As long as everything outside of datagen implements the "actual" traits, then we can safely add the blanket impl in datagen, right?

Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: No, because the cldr/uprops/etc providers need to make a Multi provider of trait objects.

Comment (sffc) [reply provider/core/src/datagen/mod.rs:26]: I thought we could implement a trait on a foreign type so long as that impl is in the same crate as the definition of the trait. We could implement Omnibus for Multi inside of the datagen crate, no?

Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: @sffc we could; the problem is the trait objects involved when constructing the Multi in provider_cldr (etc)

the trait needs to exist then, otherwise the fact that it is implemented on each subprovider gets erased

Comment (sffc) [reply provider/core/src/datagen/mod.rs:26]: I see. I didn't realize we had a trait object at the library boundary. This is a recent change from @robertbastian's refactoring work.

I think a better design is to go back to concrete types at the library boundary.

To be clear, I think we can/should do this in a follow-up PR.

Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: I'm unsure how we can do concrete types at the boundary with MultiBlob unless we do some kind of tuple implementation (since Rust doesn't have variadic generics)

We actually don't need this trait at all if not for this boundary :smile:, it purely exists to make that trait object work.

Comment (sffc) [reply provider/core/src/datagen/mod.rs:26]: Previously we exported each provider on its own (NumbersProvider, TimeZonesProvider, CalendarsProvider, ...). Maybe we should go back to that model.

Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: Ah, gotcha. It _is_ convenient to be able to have the list of all sub-providers in one place though

Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: (perhaps we could have that list be in the datagen _crate_)

Comment (sffc) [reply provider/core/src/datagen/mod.rs:26]: > (perhaps we could have that list be in the datagen _crate_)

This would be fine with me. Alternatively: https://github.com/unicode-org/icu4x/pull/1684

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 1 above)
Comment (robertbastian) [reply tools/datagen/src/bin/verify-zero-copy.rs:82]: The difference with `ALL_KEYS` over `TEST_KEYS` is that in line 220 we wouldn't ignore key errors.

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (robertbastian) [reply provider/core/src/datagen/mod.rs:26]: That is a cool solution, can you incorporate #1684 ?

File: provider/core/src/dynutil.rs (hunk 8 above)
Comment (robertbastian) [reply provider/core/src/dynutil.rs:207]: What ResourceProvider stuff?

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: tools/datagen/src/bin/verify-zero-copy.rs (hunk 1 above)
Comment (Manishearth) [reply tools/datagen/src/bin/verify-zero-copy.rs:82]: That means all-keys would always error, datagen doesn't have proper support for this either. I'm only adding features that'll actually be useful.

File: provider/core/src/datagen/mod.rs (hunk 7 above)
Comment (Manishearth) [reply provider/core/src/datagen/mod.rs:26]: It's not sufficient in and of itself without further work: you either need to name the really complex type over the boundary or you need to do a bunch of manual passthrough implementations. I'd rather land this first, I don't consider this additional trait to be that much overhead.

**Review by robertbastian — APPROVED** [review]
